### ISSUE TITLE
Port opens but can't read data

### ISSUE THREAD

**Issue by Zopenzop** [FengChendian/serial_port_win32#14]
I have confirmed that the connection between my laptop and my serial device is correct and working. But when using this package, I am able to get the port, open it but it never reads data. I have no idea why it is so. The baud rate and everything is just as it should be, and I am using the example project made by you. What could be the reason?

**Comment by nirmalaranawat** [comment]
I am facing similar issue. I am able to open and write to ports but not able to read the data. Facing the same issue with example code provided.

**Comment by FengChendian** [comment]
How much data do you read at a time?

I just translated C++ code to dart interface. But win32 package is a litter different from Win32 API.
The num of read bytes requires num of bytes you send.

If you send 4 bytes data, but you read 5 bytes. Maybe you can not get data.

**Comment by nirmalaranawat** [comment]
Thank you for response.  In my application I do not know before hand the number of bytes I will be receiving. 
I tried sending a fixed length of data and read it on the app, but that did not happen either.

**Comment by Zopenzop** [comment]
@FengChendian At a time I receive 8-10 bytes of data. Also, if ReadIntervalTimeout, ReadTotalTimeoutConstant or ReadTotalTimeoutMultiplier are not suitable for my project, should I fork this repo and get rid of them?

**Comment by FengChendian** [comment]
@nirmalaranawat @Zopenzop I can get data correctly even though data size is larger than transmited by MCU. Do you have any small code example?
<img width="1002" alt="image" src="https://user-images.githubusercontent.com/37545115/173020671-04e9370d-a00e-495f-99ae-3e840afe2d86.png">

**Comment by FengChendian** [comment]
> @FengChendian At a time I receive 8-10 bytes of data. Also, if ReadIntervalTimeout, ReadTotalTimeoutConstant or ReadTotalTimeoutMultiplier are not suitable for my project, should I fork this repo and get rid of them?

You can set them when create serial instance. Like
```dart
port = SerialPort(ports[0], openNow: false, ReadIntervalTimeout: 1, ReadTotalTimeoutConstant: 2);
```

**Comment by lesslefou** [comment]
Hello, 
I'm dealing with the same error and I don't know how to resolve it. I'm a new flutter coder. 

I use the example code and I try to add all the parameters settings for the port but still nothing.  How do you resolve this problem ? 
![image](https://user-images.githubusercontent.com/51909376/182626902-624953ba-90a2-4587-99a4-f906c389611d.png)


Here are the data I receive periodically by my sensor. 
![image](https://user-images.githubusercontent.com/51909376/182626443-bd79fdc8-d91a-46db-8ab1-5ea6f15cf066.png)

Thanks for your help.

**Comment by lesslefou** [comment]
> Is there any solution for this? mine won't work either.

Personnaly I didn't realise to make the code work so I change to the following librairy : flutter_libserialport.

**Comment by wachidsusilo** [comment]
@lesslefou Does flutter_libserialport works for you? 
I initially came to that library, but I thought this platform specific lib would be more stable on windows.
may be i should go back to the initial plan.

**Comment by lesslefou** [comment]
Well I succeed to do what I want with this librairy. I just have to read data with a terminal before launching my application. I don't know why if I don't do that I receive strange data.

**Comment by FengChendian** [comment]
@wachidsusilo @lesslefou Do you have a whole project and I/O data example? I don't know why some people can't read data correctly. And other people can read proper data. It's confused.

Maybe due to typecasting:
```dart
 var u8l = lpBuffer.cast<Uint8>().asTypedList(_bytesRead.value);
```

**Comment by wachidsusilo** [comment]
@FengChendian may be the problem was in dart win32 API, as i used `flutter_libserialport` library and the problem still persist.
It's working though, but i need to open the `SerialPort` on **another program** first. 

Actually, while debugging i found interesting behaviour:
1. Open Serial -> success
2. Read Data -> nothing shown
3. Write Data -> nothing sent
4. Close Serial -> success
5. Open Serial on another program -> The previously written data was sent immediately

May be this was due to `NON_OVERLAPPED_IO` thing?
Currently i give up on this. I've changed the source file, so i cannot give you the example.

I create my own DLL using `ILCompiler` and `NativeAOT` on C# (DotNet). I wrote my own `ffi` implementation and it works as expected now.

By the way, you did a great job @FengChendian. The problem was on me. I was doing great on `Kotlin`, i'm regretting choosing this path (flutter) 😂

**Comment by FengChendian** [comment]
Maybe due to `OVERLAPPED_IO` or win32 API.

I use `OVERLAPPED_IO` to get data in async mode. Otherwise the function  `WaitForSingleObject` will enter blocking mode.Then UI will be freezed.

(⊙︿⊙) But if win32 API causes this problem, I think there is nothing I can do.

**Comment by Zopenzop** [comment]
@FengChendian I don't think it's the Win32 api. Anyways, I'll be back to working on my project once again and I'll revert back with how it turns out to work.

**Comment by maiducgiang** [comment]
Hi,
I can't read the complete input message,
in image first, I read fail,
in image second, I read by app other
![image](https://user-images.githubusercontent.com/43919583/203962737-f93975d2-c0cc-4bc7-bf6e-0510de19de7d.png)
![image](https://user-images.githubusercontent.com/43919583/203962859-02ceeba1-16c7-4647-b64e-dfff6f450461.png)

**Comment by maiducgiang** [comment]
that is my setup,
sorry my english is bad
![image](https://user-images.githubusercontent.com/43919583/203963338-c815ae9d-ccf7-4c8e-bb01-345a789f1199.png)

**Comment by FengChendian** [comment]
> Hi, I can't read the complete input message, in image first, I read fail, in image second, I read by app other ![image](https://user-images.githubusercontent.com/43919583/203962737-f93975d2-c0cc-4bc7-bf6e-0510de19de7d.png) ![image](https://user-images.githubusercontent.com/43919583/203962859-02ceeba1-16c7-4647-b64e-dfff6f450461.png)

Does your message have '\0' ending? According to discussion, '\0' ending is important for message to be truncated by ReadFile Function. 
And readBytesSize == messageSize(include '\0') is better.

**Comment by maiducgiang** [comment]
what can i do?

**Comment by FengChendian** [comment]
> what can i do?

If you send 10 char, read 10bytes, not 100.
And try add '\n' or '\0' in your message. 
In C
```c
char message[5] = '1234'; // it is actually '1234\0'

uint8_t data[5] = {1, 2, 3, 4, 5}; // not ending with  '\0'
```
if it doesn't work, please try other package implemented by native C/C++ language.

**Comment by arcral** [comment]
i also can't read any data.  i think the read function has some serious bugs

**Comment by FengChendian** [comment]
> i also can't read any data. i think the read function has some serious bugs

If you use `readBytesOnce` function, can you read anything?
Or do you have any code to reproduce this bug?

There are some reports that the read function doesn't work properly, but I can't reproduce the bug on my computer.
The biggest difference between the read and write functions is that read uses async and stream, so I don't know if stream is causing this bug or win32 API init failed.

**Comment by FengChendian** [comment]
I have confirmed that `ClearCommError` and `PurgeComm` causes the read bug. `ClearCommError` will return wrong  `cbInQue` in some systems if `PurgeComm` has been executed.

Because `PurgeComm` may be processed after rx ready due to CPU delay. (It should be processed before rx ready)

I have removed the stage of purge com in read function and tested it using ESP32. I think I have fixed this bug in version 1.1.0

**Comment by FengChendian** [comment]
I have confirmed that `ClearCommError` and `PurgeComm` causes the read bug. `ClearCommError` will return wrong  `cbInQue` in some systems if `PurgeComm` has been executed.

Because `PurgeComm` may be processed after rx ready due to CPU delay. (It should be processed before rx ready)

I have removed the stage of purge com in read function and tested it using ESP32. I think I have fixed this bug in version 1.1.0
